### DISCUSSION TITLE
Make `empty` and `nempty` work for Array types

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#24889] [Feature Requests]
### Discussed in https://github.com/directus/directus/discussions/18039

<div type='discussions-op-text'>

<sup>Originally posted by **soulseekah** April  3, 2023</sup>
### Summary

There is currently no way to check if an array is empty or not.

`_empty` does not work as per https://github.com/directus/directus/issues/18028

while `"_eq": []` is not supported

### Basic Example

Create a flow, Read step that returns [] when nothing is found, Condition that checks `{"$last": {"_empty": true}}` or `{"$last":{"_eq": []}}`. First option expects `''`, `false` or `null`. Second option errors out as unsupported operand.

### Motivation

Checking if an array is empty or not is a quite important and foundational construct to have available in any system.

### Detailed Design

`_empty` and `_nempty` should treat an empty `[]` as falsy, despite them being always truthy.

### Must Have:

- `_empty` checks for `length` of 0 on array typed values
- `_nempty` checks for `length` of >= 1 on array typed values

### Should Have:

n/a

### Could Have:

n/a

### Won't Have:

n/a

### Drawbacks

There are no known drawbacks to supporting arrays in `_empty` and `_nempty`.

### Alternatives

Writing a custom script that checks `.length` is a usable approach, before applying a condition.

![image](https://user-images.githubusercontent.com/685880/229561163-2379084a-1527-48b4-a771-099ab7aeb86c.png)

### Adoption Strategy

`_empty` will start supporting arrays. A notice will be posted on the release notes to notify people of the possible breaking change

### Unresolved Questions

n/a</div>

**Comment by shipsides-sparknel** [comment]
#15504  ([] not being considered as null for field validation) is tangentially related to this, which was attempted to be fixed but instead caused #16291. is finding a solution for this part of the scope of this?

**Comment by jekuer** [comment]
The mentioned alternative does not work well when filtering within a (for example) delete flow step, where it is about a lot of items. Of course, you can also do this via a script, but feels like unnecessarily complicated.

What you can do, right within the filter (since function parameters are now possible):
```
{
    "count(field)": {
        "_eq": 0
    }
}
```

**Maybe an alternative approach for this ticket is to update documentation instead of updating "_empty"?**

**Comment by MarvinXu** [comment]
> The mentioned alternative does not work well when filtering within a (for example) delete flow step, where it is about a lot of items. Of course, you can also do this via a script, but feels like unnecessarily complicated.
> 
> What you can do, right within the filter (since function parameters are now possible):
> 
> ```
> {
>     "count(field)": {
>         "_eq": 0
>     }
> }
> ```
> 
> **Maybe an alternative approach for this ticket is to update documentation instead of updating "_empty"?**



@jekuer how did you make type check work for this?

**Comment by jekuer** [comment]
@MarvinXu , you mean before whether it's an array or not OR regarding the type of the array element?

Regarding first: This is quite static in my setup. Not sure, but most probably you could also work with the "_or" operator. like nnull = false OR count = 0. This would fail if not given at all, null or an empty array.

Regarding latter one: I don't 😆

**Comment by MarvinXu** [comment]
@jekuer sorry, I mean Typescript error
![image](https://github.com/directus/directus/assets/7382357/5c5b5921-c2d1-42e4-a62d-e2ced7993a5b)

**Comment by jekuer** [comment]
Good catch. I don't, because I use this only at the integrated Flow-Feature of Directus. In my app, I am filtering out empty arrays after fetching the data from the Directus backend.

**Comment by useEffects** [comment]
```
_and: [
                    {
                        "count(members)": {
                            _gt: 0
                        }
                    },
                    {
                        members: {
                            _nnull: true
                        }
                    }
                ]
```
this is better

**Comment by lundjrl** [comment]
> ```
> _and: [
>                     {
>                         "count(members)": {
>                             _gt: 0
>                         }
>                     },
>                     {
>                         members: {
>                             _nnull: true
>                         }
>                     }
>                 ]
> ```
> 
> this is better

This resolved the issue for me
